libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -0,0 +1,453 @@
+                id,
+                recv,
+                sender: self.sender.clone(),
+            }),
+            Err(_) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
+            )),
+        }
+    }
+}
+
+pub struct Strand {
+    id: String,
+    sender: Sender<PmMessage>,
+    recv: Receiver<Response>,
+    timeout: Duration,
+}
+
+impl Strand {
+    pub fn request_connection(&self, endpoint: &str) -> Result<Response, ConnectionManagerError> {
+        let message = PmMessage::Request(Request {
+            from: self.id.clone(),
+            payload: Payload::AddConnection {
+                endpoint: endpoint.to_string(),
+            },
+        });
+
+        match self.sender.send_timeout(message, self.timeout.clone()) {
+            Ok(()) => (),
+            Err(SendTimeoutError::Timeout(_)) => {
+                return Err(ConnectionManagerError::SendTimeoutError(
+                    "Disconnection request timed out".into(),
+                ))
+            }
+            Err(SendTimeoutError::Disconnected(_)) => {
+                return Err(ConnectionManagerError::SendMessageError(
+                    "The connection manager is no longer running".into(),
+                ))
+            }
+        };
+
+        self.recv
+            .recv()
+            .map_err(|err| ConnectionManagerError::SendMessageError(format!("{:?}", err)))
+    }
+
+    pub fn disconnect(&self) -> Result<(), ConnectionManagerError> {
+        let message = PmMessage::Disconnect(self.id.clone());
+        match self.sender.send_timeout(message, self.timeout.clone()) {
+            Ok(()) => Ok(()),
+            Err(SendTimeoutError::Timeout(_)) => Err(ConnectionManagerError::SendTimeoutError(
+                "Disconnection request timed out".into(),
+            )),
+            Err(SendTimeoutError::Disconnected(_)) => {
+                Err(ConnectionManagerError::SendMessageError(
+                    "The connection manager is no longer running".into(),
+                ))
+            }
+        }
+    }
+}
+
+pub struct NotificationHandler {
+    id: String,
+    sender: Sender<PmMessage>,
+    recv: Receiver<Notification>,
+}
+
+impl NotificationHandler {
+    pub fn listen(&self) -> Result<Option<Notification>, ConnectionManagerError> {
+        match self.recv.try_recv() {
+            Ok(notification) => Ok(Some(notification)),
+            Err(TryRecvError::Disconnected) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
+            )),
+            Err(TryRecvError::Empty) => Ok(None),
+        }
+    }
+
+    pub fn unsubscribe(&self) -> Result<(), ConnectionManagerError> {
+        let message = PmMessage::UnSubscribe(self.id.clone());
+        match self.sender.try_send(message) {
+            Ok(()) => Ok(()),
+            Err(_) => Err(ConnectionManagerError::SendTimeoutError(
```

Comment:
Why is this considered a timeout? The other possibility is that the receiver is gone.